src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -176,52 +243,162 @@ async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)>
+            int position,
+            SemanticModel semanticModel,
+            ITypeSymbol receiverTypeSymbol,
+            MultiDictionary<ITypeSymbol, IMethodSymbol> matchingMethodSymbols,
+            ArrayBuilder<SerializableImportCompletionItem> builder,
+            Dictionary<INamespaceSymbol, string> stringCache,
+            Dictionary<ITypeSymbol, bool> checkedReceiverTypes,
+            CancellationToken cancellationToken)
+        {
+            // Matching extension method symbols are grouped based on their receiver type.
+            foreach (var (declaredReceiverType, methodSymbols) in matchingMethodSymbols)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var declaredReceiverTypeInCurrentCompilation = SymbolFinder.FindSimilarSymbols(declaredReceiverType, semanticModel.Compilation).FirstOrDefault();
+                if (declaredReceiverTypeInCurrentCompilation == null)
+                {
+                    // Bug: https://github.com/dotnet/roslyn/issues/45404
+                    // SymbolFinder.FindSimilarSymbols would fail if current and referenced compilation targeting different frameworks say net472 and netstandard respectively.
+                    // Here's SymbolKey for System.String from those two framework as an example:
+                    //
+                    //  {1 (D "String" (N "System" 0 (N "" 0 (U (S "netstandard" 4) 3) 2) 1) 0 0 (% 0) 0)}
+                    //  {1 (D "String"(N "System" 0(N "" 0(U(S "mscorlib" 4) 3) 2) 1) 0 0(% 0) 0)}
+                    //
+                    // Also we don't use the "ignoreAssemblyKey" option for SymbolKey resolution because its perfermance doesn't meet our requirement.
+                    continue;
+                }
+
+                if (checkedReceiverTypes.TryGetValue(declaredReceiverTypeInCurrentCompilation, out var cachedResult) && !cachedResult)
+                {
+                    // If we already checked an extension method with same receiver type before, and we know it can't be applied
+                    // to the receiverTypeSymbol, then no need to proceed methods from this group..
+                    continue;
+                }
+
+                // This is also affected by the symbol resolving issue mentioned above, which means in case referenced projects
+                // are targeting different framework, we will miss extension methods with any framework type in their signature from those projects.
+                var methodsInCurrentCompilation = methodSymbols.Select(s => SymbolFinder.FindSimilarSymbols(s, semanticModel.Compilation).FirstOrDefault()).WhereNotNull();
+
+                var enumerator = methodsInCurrentCompilation.GetEnumerator();
+                if (!enumerator.MoveNext())
+                {
+                    continue;
+                }
+
+                var methodInCurrentCompilation = enumerator.Current;
```

Comment:
can this just be a foreach, where you lazily do the processing of the reducedMethodSymbol the first time you hit it?

Reply:
Maybe, but I think using enumerator directly result sin cleaner code in this case.

Reply:
The main reason for this is to avoid calling `SymbolFinder.FindSimilarSymbols` unnecessarily.

Reply:
changed to loop